database/src/tables.rs

```diff
@@ -4,26 +4,271 @@
 
 //---------------------------------------------------------------------------------------------------- Import
 use crate::{
+    database::{DatabaseIter, DatabaseRo, DatabaseRw},
     table::Table,
     types::{
-        Amount, AmountIndex, AmountIndices, BlockBlob, BlockHash, BlockHeight, BlockInfoV1,
-        BlockInfoV2, BlockInfoV3, KeyImage, Output, PreRctOutputId, PrunableBlob, PrunableHash,
-        PrunedBlob, RctOutput, TxHash, TxId, UnlockTime,
+        Amount, AmountIndex, AmountIndices, BlockBlob, BlockHash, BlockHeight, BlockInfo, KeyImage,
+        Output, PreRctOutputId, PrunableBlob, PrunableHash, PrunedBlob, RctOutput, TxBlob, TxHash,
+        TxId, UnlockTime,
     },
 };
 
-//---------------------------------------------------------------------------------------------------- Tables
+//---------------------------------------------------------------------------------------------------- Sealed
 /// Private module, should not be accessible outside this crate.
-///
-/// Used to block outsiders implementing [`Table`].
-/// All [`Table`] types must also implement [`Sealed`].
 pub(super) mod private {
     /// Private sealed trait.
     ///
     /// Cannot be implemented outside this crate.
     pub trait Sealed {}
 }
 
+//---------------------------------------------------------------------------------------------------- `trait Tables[Mut]`
+/// Creates:
+/// - `pub trait Tables`
+/// - `pub trait TablesMut`
+/// - Blanket implementation for `(tuples, containing, all, open, database, tables, ...)`
+///
+/// For why this exists, see: <https://github.com/Cuprate/cuprate/pull/102#pullrequestreview-1978348871>.
+macro_rules! define_trait_tables {
```

Comment:
## `EnvInner` / `trait Tables{Mut,Iter}`

- Added `trait Tables{Mut,Iter}` as discussed before
- Added `fn open_tables()`, `fn open_tables_mut()` to return those `impl Tables{Mut,Iter}`

`trait TablesIter` was added so we can do `trait TablesMut: Tables` so
our `impl TablesMut` objects have access to multiple table reads too.